content/blog/2025-04-09-introducing-bloom-filters.md

```diff
@@ -0,0 +1,154 @@
+This will require variable memory depending on the false positive rate. In all cases (even stricter false positive rates), we can see there is a significant memory optimization compared to using the `SET` data type.
+
+| Number of Bloom Filters | Capacity | FP Rate | FP Rate Description  | Total Used Memory (GB) | Memory Saved % compared to SETS |
+|-------------------------|----------|---------|----------------------|------------------------|----------------------------------|
+| 500                     | 5000000  | 0.01    | One in every 100      | 2.9                    | **98.08%** |
+| 500                     | 5000000  | 0.001   | One in every 1K       | 4.9                    | **96.80%** |
+| 500                     | 5000000  | 0.00001 | One in every 100K     | 7.8                    | **94.88%** |
+| 500                     | 5000000  | 0.0000002| One in every 5M      | 9.8                    | **93.60%** |
+
+In this example, we are able to benefit from 93% - 98% savings in memory usage when using Bloom Filters compared to the SET data type. Depending on your workload, you can expect similar results.
+
+![SET vs Bloom Filter Memory Usage Comparison](/assets/media/pictures/bloomfilter_memusage.png)
+
+## Large Bloom Filters and Recommendations
+
+To ensure optimal server performance during serialization and deserialization of bloom filters, we have added validation on the memory usage per object.
+The default memory usage limit of a bloom filter is defined by the `BF.BLOOM-MEMORY-USAGE-LIMIT` module configuration which has a default value of 128 MB.
+However, the value can be tuned using the configuration above.
+
+The implication of the memory limit is that operations involving bloom filter creations or scaling out that result in a bloom filter with overall memory usage over the limit will be rejected. Example:
+```
+127.0.0.1:6379> BF.ADD ad1_filter user1
+(error) ERR operation exceeds bloom object memory limit
+```
+This poses an issue to users where their scalable bloom filters can reach the memory limit after some number of days of data population and it starts failing scale outs during the insertion of unique items.
+
+As a solution, to help users understand at what capacity their bloom filter will hit the memory limit, valkey-bloom has two options.
+These are useful to check beforehand to ensure that your bloom filter will not fail scale outs or creations later on as part of your workload.
+
+1. Perform a memory check prior to bloom filter creation
+
+We can use the `VALIDATESCALETO` option of the `BF.INSERT` command to perform a validation whether the filter is within the memory limit.
+If it is not within the limits, the command is rejected. In the example below, we see that filter1 cannot scale out and reach the capacity of 26214301 due to the memory limit. However, it can scale out and reach a capacity of 26214300.
```

Comment:
```suggestion
If it is not within the limits, the command will return an error. In the example below, we see that filter1 cannot scale out and reach the capacity of 26214301 due to the memory limit. However, it can scale out and reach a capacity of 26214300.
```